src/lib/checksum.lua

```diff
@@ -68,6 +68,62 @@ function verify_packet (buf, len)
    return ipsum(buf+headersize, len-headersize, initial) == 0
 end
 
+local
+function prepare_packet_l4( buf, len, csum_start, csum_off)
+
+  local hwbuf =  ffi.cast('uint16_t*', buf)
+
+  local pheader = C.pseudo_header_initial(buf, len)
+  if band(pheader, 0xFFFF0000) == 0 then
+    hwbuf[(csum_start+csum_off)/2] = C.htons(band(pheader, 0x0000FFFF))
+  else
+    csum_start, csum_off = nil, nil
+  end
+
+  return csum_start, csum_off
+end
+
+function prepare_packet4( buf, len)
+
+  local hwbuf =  ffi.cast('uint16_t*', buf)
+  local proto = buf[9];
+
+  local csum_start = lshift(band(buf[0], 0x0F),2)
+  local csum_off
+
+  -- Update the IPv4 checksum (use in-place pseudoheader, by setting it to 0)
+  hwbuf[5] = 0;
+  hwbuf[5] = C.htons(C.cksum_generic(buf, csum_start, 0));
+
+  -- TCP
+  if proto == 6 then
+    csum_off = 16
+  -- UDP
+  elseif proto == 17 then
+    csum_off = 6
+  end
+
+  return prepare_packet_l4( buf, len, csum_start, csum_off)
+end
+
+function prepare_packet6( buf, len)
+  local hwbuf =  ffi.cast('uint16_t*', buf)
+  local proto = buf[6];
+
+  local csum_start = 40
+  local csum_off
+
+  -- TCP
+  if proto == 6 then
+    csum_off = 16
+  -- UDP
+  elseif proto == 17 then
+    csum_off = 6
+  end
+
+  return prepare_packet_l4( buf, len, csum_start, csum_off)
+end
+
```

Comment:
The checksum functionality is now optional.
The functions are meant to calculate partial checksums, and for me, this is where they belong.
For sure they are not virtio specific.